#parameters file

alts={
	"Spec_Rg":"analysed_region",
	"Deletion":"del",
	"Duplication":"dup",
	"Inversion":"inv",
	"Balanced_translocation":"t",
	"Insertion":"ins",
	"Unbalanced_translocation":"Unb_t",
}

hg19={
	"version":"hg19",
	"is_hit_in":False,
	"infertility":"data/infertility",
	"acmg":"data/ACMG_59_Genes",
	"band":"data/cytoband_hg19.bed",
	"at1":['start_position', 'end_position','external_gene_name', 'strand',  'gene_biotype', 'ensembl_gene_id', 'external_synonym'],
	"els":["Gene start (bp)", "Gene end (bp)","Gene name", "Gene type", "Gene stable ID", "Strand", "Gene Synonym"],
	"at2":['ensembl_gene_id','ensembl_transcript_id', 'exon_chrom_start','exon_chrom_end','gene_biotype', 'transcript_biotype',"transcript_start", "transcript_end"],
	"at3":['external_transcript_name','external_gene_name','ensembl_transcript_id', "strand"],
	"els2":["Gene stable ID", "Transcript type", "Gene type","Exon region start (bp)", "Exon region end (bp)", "Transcript stable ID", "Transcript start (bp)", "Transcript end (bp)"],
	"els3":["Transcript stable ID","Transcript name", "Gene name", "Strand"],
	"chrsz":"data/hg19_sizes",
	"snps":["data/trait_and_genes", "data/ontology_correspondance", "data/SNP_table_hg38_13_3_2019"],
	"DDG2P":"data/DDG2P_21_2_2019.csv",
	"clingen":"data/clingen2",
	"inh":"data/gen2phen2inh_julh2020",
	"oe":"data/oe_alt",
	"cen":"data/centromere_hg19",
	"panel":"data/panel_list",
	"hi":"data/hi_triplo_hg19",
	"at15":['ensembl_gene_id','external_gene_id'],
	"els15":["Ensembl Gene ID", "Associated Gene Name"],
	"onco":"data/tonco_bd",
	"tonco":"data/tiny_onco",
	"gtex":"data/top3gtext",
	"genehancer":"data/GeneHancer_cluster_interactions",
	"animals":{"wormBaseId":["data/wormbase_formated_bd", "https://wormbase.org/species/c_elegans/gene/"],
	"zfinId":["data/zfin_formated_bd", "http://zfin.org/"],
	"rgdId":["data/rat_formated_bd", "https://rgd.mcw.edu/rgdweb/report/gene/main.html?id="],
	"flyBaseId":["data/flybase_formated_bd", "http://flybase.org/reports/"],
	"mgiId":["data/mgi_formated_bd", "http://www.informatics.jax.org/marker/"]},
	"for_y":{"sz":59373566, "Consensus":1786131, "IMR90 (fibroblasts)":824472, "LCL (GM12878)":8359467, "hESC":1058135, "A549": 1411623, "Aorta":1690650, "Cortex":1337331, "Bladder":1672068, "Lung":1435346, "HUVEC":923778, "K562":867120},
	"coments":{"A":"GeneCard link of protein coding genes, lincRNA, expressed genomic elements with Gtex data and disrupted elements. Virtual panel genes with a high level of evidence of disease causation, letter coded associated phenotypes and level of evidence (*) are indicated.",
        "B":"Medically actionable genes in which pathogenic or likely pathogenic variants are reported as incidental or secondary findings.",
        "C":"Region of the genomic element affected by the rearrangement; Genome strand: sense strand (SS) or antisense strand (AS)",
        "D":"OMIM gene ID with hyperlink to the respective OMIM webpage.",
        "E":"Haploinsuficiency index (HI%) according to Decipher and the triplosensitivity score (Triplo) according to ClinGen.",
        "F":"Probability of Loss-of-function Intolerance (pLi); Ratio of the observed / expected number of loss-of-function variants in that gene (o/e), and respective confidence interval, according to gnomAD.",
        "G":"The Uniprot search for the specific gene in Human.",
        "H":"Gene Function description according to OMIM. Due to the nature of the function description and te limitation of characters by cell, this description may be incomplete.",
        "I":"Top 3 more expressed cells/tissues in between 54 types, according to GTEx. Expression is indicated in Transcripts Per Million (TPM) for the Highest Median Expression, Total Median Expression and Average Expression.",
        "J":"Location of regulatory elements associated with the respective genes, according to GeneHancer.",
        "K":"Chromatin loops form a basic unit of interphase nuclear organization, with chromatin loop anchor points providing contacts between regulatory regions and promoters. The chromatin loops presented are according to the cell line chosen.The genomic elements forming a loop and their respective coordinates are separated by an underscore.",
        "L":"Pathologies associated with the reference gene in OMIM, DDG2P and/or ClinGendatabases",
        "M":"OMIM Phenotype ID, followed by inheritance type.",
        "N":"Development Disorder Genotype - Phenotype Database (DDG2P) classification of the phenotype.",
        "O":"ClinGen database classification of the phenotype.",
        "P":"Phenotype similarity analysis (Result of the overlap between the inputted HPO terms and the gene-associated HPO terms). The overlap score is showed, followed by the p-value that shows the probability of the score to be obtained by chance, and the maximum score that can be obtained with the inputed terms.",
        "Q":"Type of cancer and respective number of cases, according to Atlas of Genetics and Cytogenetics in Oncology and Haematology and Mitelman Database of Chromosome Aberrations and Gene Fusions in Cancer.",
        "R":"Fusion genes followed by their origin rearrangement, according to Atlas of Genetics and Cytogenetics in Oncology and Haematology and Mitelman Database of Chromosome Aberrations and Gene Fusions in Cancer.",
        "S":"Curated disease associations based on experimental data from Wormbase, and the direct link to Uniprot search of the gene in C.elegans.",
        "T":"Curated disease associations based on experimental data from Flybase, , and the direct link to Uniprot search of the gene in drosophila.",
        "U":"Curated disease associations based on experimental data from MGI, and the direct link to Uniprot search of the gene in mouse.",
        "V":"Curated disease associations based on experimental data from RGD, and the direct link to Uniprot search of the gene in rat.",
        "W":"Curated disease associations based on experimental data from Zfin, and the direct link to Uniprot search of the gene in zebrafish.",
        "X":"Infertility disorders associated to the genes.",
        "Y":"Genome-wide association studies (GWAS), SNPs grouped according to genomic regions, associated genetic traits or phenotypic features with the respective p-value, are presented.Data is according to GWAS catalog.",
        "Z":"Hyperlink to PubMed search with the gene name.",
        "AA":"The best overlap hits, their clinical classification and frequency, in each tested database."
	},"oes":{},
	"dic_ref":{"1":"NC_000001.10", "2":"NC_000002.11", "3":"NC_000003.11", "4":"NC_000004.11", "5":"NC_000005.9", "6":"NC_000006.11", "7":"NC_000007.13", "8":"NC_000008.10", "9":"NC_000009.11", "10":"NC_000010.10", "11":"NC_000011.9", "12":"NC_000012.11", "13":"NC_000013.10", "14":"NC_000014.8", "15":"NC_000015.9", "16":"NC_000016.9", "17":"NC_000017.10", "18":"NC_000018.9", "19":"NC_000019.9", "20":"NC_000020.10", "21":"NC_000021.8", "22":"NC_000022.10", "X":"NC_000023.10", "Y":"NC_000024.9"}
	}

hg19_lines={
	"IMR90":["IMR90 fibroblasts", "data/hg19_tads/IMR90_Lieberman-raw_TADs.txt", "Lieberman et al., 2009","data/hg19_lims/imr90_hg19_lims.txt","data/hg19_loops/imr90_hg19_loops"],#tad, tadfile, ref, li
	"LCL":["LCL (GM12878)", "data/hg19_tads/GM12878_Lieberman-raw_TADs.txt", "Lieberman et al., 2009","data/hg19_lims/GM12878_hg19_lims.txt","data/hg19_loops/gm12878_hg19_loops"],
	"hesc":["hESC", "data/hg19_tads/H1-ESC_Dixon2015-raw_TADs.txt", "Dixon et al., 2015","data/hg19_lims/hesc_hg19_lims.txt","data/hg19_loops/hesc_hg19_loops"],
	"consensus":["Consensus", "data/hg19_tads/hg19_consensusTADslifei2019.bed", "Lifei et al., 2019","data/hg19_lims/consensus_hg19_lims.txt","data/hg19_loops/none"],
	"a549":["A549", "data/hg19_tads/A549_raw-rep1_TADs.txt", "Encode, 2016","data/hg19_lims/A549_hg19_lims.txt","data/hg19_loops/A549_hg19_loops"],
	"aorta":["Aorta", "data/hg19_tads/Aorta_STL002_Leung2015-raw_TADs.txt", "Leung et al., 2015","data/hg19_lims/Aorta_hg19_lims.txt","data/hg19_loops/aorta_hg19_loops"],
	"cortex":["Cortex", "data/hg19_tads/Cortex_DLPFC_Donor-CO-raw_TADs.txt", "Schmmit et al., 2016","data/hg19_lims/Cortex_hg19_lims.txt","data/hg19_loops/cortex_hg19_loops"],
	"blader":["Bladder", "data/hg19_tads/Bladder_Donor-BL1-raw_TADs.txt", "Schmmit et al., 2016","data/hg19_lims/Bladder_hg19_lims.txt","data/hg19_loops/bladder_hg19_loops"],
	"lung":["Lung", "data/hg19_tads/Lung_Donor-LG1-raw_TADs.txt", "Schmmit et al., 2016","data/hg19_lims/lung_hg19_lims.txt","data/hg19_loops/lung_hg19_loops"],
	"huvec":["HUVEC", "data/hg19_tads/HUVEC_Lieberman-raw_TADs.txt", "Lieberman et al., 2009","data/hg19_lims/HUVEC_hg19_lims.txt","data/hg19_loops/huvec_hg19_loops"],
	"k562":["K562", "data/hg19_tads/K562_Lieberman-raw_TADs.txt", "Rao et al., 2014","data/hg19_lims/K562_hg19_lims.txt","data/hg19_loops/k562_hg19_loops"]}

hg38={
	"version":"hg38",
	"oes":{},
	"is_hit_in":False,
	"infertility":"data/infertility",
	"acmg":"data/ACMG_59_Genes",
	"at1":['start_position', 'end_position','external_gene_name', 'strand',  'gene_biotype', 'ensembl_gene_id', 'external_synonym'],
	"els":["Gene start (bp)", "Gene end (bp)","Gene name", "Gene type", "Gene stable ID", "Strand", "Gene Synonym"],
	"band":"data/cytoband_hg38.bed",
	"chrsz":"data/hg38_sizes",
	"cen":"data/centromere_hg38",
	"for_y":{"sz":57227415,"Consensus":1796527, "IMR90 (fibroblasts)":858229, "LCL (GM12878)":815201, "hESC":1087767, "A549": 1405466, "Aorta":1679636, "Cortex":1736794, "Bladder":1799711, "Lung":1754399, "HUVEC":1072369, "K562":842166},
	"snps":["data/trait_and_genes", "data/ontology_correspondance", "data/SNP_table_hg38_13_3_2019"],
	"at2":['ensembl_gene_id','ensembl_transcript_id', 'exon_chrom_start','exon_chrom_end','gene_biotype', 'transcript_biotype',"transcript_start", "transcript_end"],
	"at3":['external_transcript_name','external_gene_name','ensembl_transcript_id', "strand"],
	"els2":["Gene stable ID", "Transcript type", "Gene type","Exon region start (bp)", "Exon region end (bp)", "Transcript stable ID", "Transcript start (bp)", "Transcript end (bp)"],
	"els3":["Transcript stable ID","Transcript name", "Gene name", "Strand"],
	"DDG2P":"data/DDG2P_21_2_2019.csv",
	"clingen":"data/clingen2",
	"inh":"data/gen2phen2inh_julh2020",
	"panel":"data/panel_list",
	"oe":"data/oe_alt",
	"tonco":"data/tiny_onco",
	"gtex":"data/top3gtext",
	"hi":"data/hi_triplo_hg38",
	"at15":['ensembl_gene_id','external_gene_name'],
	"els15":["Gene stable ID", "Gene name"],
	"onco":"data/tonco_bd",
	"genehancer":"data/GeneHancer_cluster_interactions",
	"animals":{"wormBaseId":["data/wormbase_formated_bd", "https://wormbase.org/species/c_elegans/gene/"],
	"zfinId":["data/zfin_formated_bd", "http://zfin.org/"],
	"rgdId":["data/rat_formated_bd", "https://rgd.mcw.edu/rgdweb/report/gene/main.html?id="],
	"flyBaseId":["data/flybase_formated_bd", "http://flybase.org/reports/"],
	"mgiId":["data/mgi_formated_bd", "http://www.informatics.jax.org/marker/"]},
	"coments":{"A":"GeneCard link of protein coding genes, lincRNA, expressed genomic elements with Gtex data and disrupted elements. Virtual panel genes with a high level of evidence of disease causation, letter coded associated phenotypes and level of evidence (*) are indicated.",
	"B":"Medically actionable genes in which pathogenic or likely pathogenic variants are reported as incidental or secondary findings.",
	"C":"Region of the genomic element affected by the rearrangement; Genome strand: sense strand (SS) or antisense strand (AS)",
	"D":"OMIM gene ID with hyperlink to the respective OMIM webpage.",
	"E":"Haploinsuficiency index (HI%) according to Decipher and the triplosensitivity score (Triplo) according to ClinGen.",
	"F":"Probability of Loss-of-function Intolerance (pLi); Ratio of the observed / expected number of loss-of-function variants in that gene (o/e), and respective confidence interval, according to gnomAD.",
	"G":"The Uniprot search for the specific gene in Human.",
	"H":"Gene Function description according to OMIM. Due to the nature of the function description and te limitation of characters by cell, this description may be incomplete.",
	"I":"Top 3 more expressed cells/tissues in between 54 types, according to GTEx. Expression is indicated in Transcripts Per Million (TPM) for the Highest Median Expression, Total Median Expression and Average Expression.",
	"J":"Location of regulatory elements associated with the respective genes, according to GeneHancer.",
	"K":"Chromatin loops form a basic unit of interphase nuclear organization, with chromatin loop anchor points providing contacts between regulatory regions and promoters. The chromatin loops presented are according to the cell line chosen.The genomic elements forming a loop and their respective coordinates are separated by an underscore.",
	"L":"Pathologies associated with the reference gene in OMIM, DDG2P and/or ClinGendatabases",
	"M":"OMIM Phenotype ID, followed by inheritance type.",
	"N":"Development Disorder Genotype - Phenotype Database (DDG2P) classification of the phenotype.",
	"O":"ClinGen database classification of the phenotype.",
	"P":"Phenotype similarity analysis (Result of the overlap between the inputted HPO terms and the gene-associated HPO terms). The overlap score is showed, followed by the p-value that shows the probability of the score to be obtained by chance, and the maximum score that can be obtained with the inputed terms.",
	"Q":"Type of cancer and respective number of cases, according to Atlas of Genetics and Cytogenetics in Oncology and Haematology and Mitelman Database of Chromosome Aberrations and Gene Fusions in Cancer.",
	"R":"Fusion genes followed by their origin rearrangement, according to Atlas of Genetics and Cytogenetics in Oncology and Haematology and Mitelman Database of Chromosome Aberrations and Gene Fusions in Cancer.",
	"S":"Curated disease associations based on experimental data from Wormbase, and the direct link to Uniprot search of the gene in C.elegans.",
	"T":"Curated disease associations based on experimental data from Flybase, and the direct link to Uniprot search of the gene in drosophila.",
	"U":"Curated disease associations based on experimental data from MGI, and the direct link to Uniprot search of the gene in mouse.",
	"V":"Curated disease associations based on experimental data from RGD, and the direct link to Uniprot search of the gene in rat.",
	"W":"Curated disease associations based on experimental data from Zfin, and the direct link to Uniprot search of the gene in zebrafish.",
	"X":"Infertility disorders associated to the genes.",
	"Y":"Genome-wide association studies (GWAS), SNPs grouped according to genomic regions, associated genetic traits or phenotypic features with the respective p-value, are presented.Data is according to GWAS catalog.",
	"Z":"Hyperlink to PubMed search with the gene name.",
	"AA":"The best overlap hits, their clinical classification and frequency, in each tested database."
	},
	"dic_ref":{"1":"NC_000001.11", "2":"NC_000002.12", "3":"NC_000003.12", "4":"NC_000004.12", "5":"NC_000005.10", "6":"NC_000006.12", "7":"NC_000007.14", "8":"NC_000008.11", "9":"NC_000009.12", "10":"NC_000010.11", "11":"NC_000011.10", "12":"NC_000012.12", "13":"NC_000013.11", "14":"NC_000014.9", "15":"NC_000015.10", "16":"NC_000016.10", "17":"NC_000017.11", "18":"NC_000018.10", "19":"NC_000019.10", "20":"NC_000020.11", "21":"NC_000021.9", "22":"NC_000022.11", "X":"NC_000023.11", "Y":"NC_000024.10"}
	}

hg38_lines={
	"IMR90":["IMR90 fibroblasts", "data/hg38_tads/IMR90_Rao_2014-raw_TADs.txt", "Rao et al. 2014","data/hg38_lims/imr90_hg38_lims.txt", "data/hg38_loops/imr90_hg38_loops" ],#tad, tadfile, ref, li
	"LCL":["LCL (GM12878)", "data/hg38_tads/GM12878_Rao_2014-raw_TADs.txt", "Rao et al. 2014","data/hg38_lims/GM12878_hg38_lims.txt", "data/hg38_loops/gm12878_hg38_loops"],
	"hesc":["hESC", "data/hg38_tads/H1-ESC_Dixon_2015-raw_TADs.txt", "Dixon et al. 2015", "data/hg38_lims/hesc_hg38_lims.txt", "data/hg38_loops/hesc_hg38_loops"],
	"consensus":["Consensus", "data/hg38_tads/consensusTADsLifei2019.txt", "Lifei et al. 2019", "data/hg38_lims/consensus_hg38_lims.txt", "data/hg38_loops/none"],
	"a549":["A549", "data/hg38_tads/A549_raw-merged_TADs.txt", "Encode 2016", "data/hg38_lims/A549_hg38_lims.txt", "data/hg38_loops/a549_hg38_loops"],
	"aorta":["Aorta", "data/hg38_tads/Aorta_STL002_Leung_2015-raw_TADs.txt", "Leung et al., 2015", "data/hg38_lims/Aorta_hg38_lims.txt", "data/hg38_loops/aorta_hg38_loops"],
	"cortex":["Cortex", "data/hg38_tads/Cortex_DLPFC_Schmitt2016-raw_TADs.txt", "Schmitt et al., 2016", "data/hg38_lims/Cortex_hg38_lims.txt", "data/hg38_loops/cortex_hg38_loops"],
	"blader":["Bladder", "data/hg38_tads/Bladder_Schmitt2016-raw_TADs.txt", "Schmitt et al., 2016", "data/hg38_lims/Bladder_hg38_lims.txt", "data/hg38_loops/bladder_hg38_loops"],
	"lung":["Lung", "data/hg38_tads/Lung_Schmitt2016-raw_TADs.txt", "Schmitt et al., 2016", "data/hg38_lims/lung_hg38_lims.txt", "data/hg38_loops/lung_hg38_loops"],
	"huvec":["HUVEC", "data/hg38_tads/HUVEC_Rao_2014-raw_TADs.txt", "Rao et al., 2014", "data/hg38_lims/huvec_hg38_lims.txt", "data/hg38_loops/huvec_hg38_loops"],
	"k562":["K562", "data/hg38_tads/K562_Rao_2014-raw_TADs.txt", "Rao et al., 2014", "data/hg38_lims/k562_hg38_lims.txt", "data/hg38_loops/k562_hg38_loops"]
	}
